### PR TITLE
[SPARK-39040][SQL] Respect NaNvl in EquivalentExpressions for expression elimination

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Respect NaNvl in EquivalentExpressions for expression elimination.

### Why are the changes needed?

For example the query will fail:
```sql
set spark.sql.ansi.enabled=true;
set spark.sql.optimizer.excludedRules=org.apache.spark.sql.catalyst.optimizer.ConstantFolding;
SELECT nanvl(1, 1/0 + 1/0);  
```
```sql
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 4.0 failed 1 times, most recent failure: Lost task 0.0 in stage 4.0 (TID 4) (10.221.98.68 executor driver): org.apache.spark.SparkArithmeticException: divide by zero. To return NULL instead, use 'try_divide'. If necessary set spark.sql.ansi.enabled to false (except for ANSI interval type) to bypass this error.
== SQL(line 1, position 17) ==
select nanvl(1 , 1/0 + 1/0)
                 ^^^    at org.apache.spark.sql.errors.QueryExecutionErrors$.divideByZeroError(QueryExecutionErrors.scala:151)
 ```
We should respect the ordering of conditional expression that always evaluate the predicate branch first, so the query above should not fail.

### Does this PR introduce _any_ user-facing change?

yes, bug fix

### How was this patch tested?

add test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @viirya @cloud-fan

File: sql/core/src/test/resources/sql-tests/inputs/ansi/conditional-functions.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+-- Tests for conditional functions
+CREATE TABLE t USING PARQUET AS SELECT c1, c2 FROM VALUES(1, 0),(2, 1) AS t(c1, c2);
+
+SELECT nanvl(c1, c1/c2 + c1/c2) FROM t;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/ansi/conditional-functions.sql:4]: so constant folding won't break it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 2]
```diff
@@ -140,6 +140,7 @@ class EquivalentExpressions {
     case i: If => i.predicate :: Nil
     case c: CaseWhen => c.children.head :: Nil
     case c: Coalesce => c.children.head :: Nil
+    case n: NaNvl => n.left :: Nil
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:143]: Can you update the comment of `childrenToRecurse` method too?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:145]: sure, add the comment

**Review by viirya — APPROVED** [review]

**Comment by ulysses-you** [comment]
thank you@viirya, any comments? @cloud-fan

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Pull request closed without merging**